README.md
![Publish Package](https://github.com/siriwatknp/react-native-video-extension/workflows/Publish%20Package/badge.svg)
![NPM](https://img.shields.io/npm/l/react-native-video-extension)


# React Native Video (Extension)

A wrapper library for [react-native-video](https://github.com/react-native-video/react-native-video) that provide great video experience for user.

- support fullscreen mode for both iOS & Android
- auto rotate video
- lightweight
- able to use with react-navigation (see example)
- icon & color configurable
- written in typescript
- works with Mux

**Currently, comes with 2 styles**

| Youtube                                                                                                         | Facebook                                                                                                        |
| --------------------------------------------------------------------------------------------------------------- | --------------------------------------------------------------------------------------------------------------- |
| ![image](https://user-images.githubusercontent.com/18292247/103453363-c7fa9800-4d0b-11eb-9ad0-48a3eb88b0f6.png) | ![image](https://user-images.githubusercontent.com/18292247/103453476-f2992080-4d0c-11eb-9b16-88ebf1c560a0.png) |

## Installation

```bash
yarn add react-native-video react-native-video-extension
```

or

```bash
npm install --save react-native-video react-native-video-extension
```

## Usage

### Basic

```tsx
import React from 'react';
import { View } from 'react-native';
import { YoutubePlayer } from 'react-native-video-extension';
// or use facebook player style
// import { FacebookPlayer } from "react-native-video-extension";

function Screen() {
  return (
    <View>
      <YoutubePlayer
        mode="auto-fit"
        source={{
          uri: 'https://stream.mux.com/Tyu80069gbkJR2uIYlz2xARq8VOl4dLg3.m3u8',
        }}
      />
    </View>
  );
}
```

See the [full code](/src/examples/BasicExample.tsx)

> Note: each screen should have only one type of player!

### More examples

- [SafeArea](https://github.com/siriwatknp/react-native-video-extension/tree/main/src/examples/SafeAreaExample.tsx)
- [ScrollView](https://github.com/siriwatknp/react-native-video-extension/tree/main/src/examples/ScrollViewExample.tsx)
- [FlatList](https://github.com/siriwatknp/react-native-video-extension/tree/main/src/examples/FlatListExample.tsx)
- [StackNavigation](https://github.com/siriwatknp/react-native-video-extension/tree/main/src/examples/StackNavigationExample.tsx)
- [MuxData](https://github.com/siriwatknp/react-native-video-extension/tree/main/src/examples/MuxDataExample.tsx)

## API

### YoutubePlayer, FacebookPlayer

|   PropName    | type                               | required |    default     | description                                                                                                                                                                                                                                 |
| :-----------: | ---------------------------------- | :------: | :------------: | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
|     mode      | "auto-fit" \| "contain"            |    ✅    |                | `auto-fit` : When rotate device or enter fullscreen manually, the video will display on the orientation that it fit the device the most <br /><br /> `contain` : In fullscreen, the video will always display the same orientation as user. |
| initialPaused | boolean                            |          |    `false`     | pause video on mount                                                                                                                                                                                                                        |
| initialMuted  | boolean                            |          |    `false`     | mute video on mount                                                                                                                                                                                                                         |
|  aspecRatio   | number \| "portrait" \| "landscape |          |   `portrait`   | the ratio of the video when it is not in fullscreen mode <br /><br /> note: landscape is `16:9`, portrait is `3:4`                                                                                                                          |
|  customIcon   | object                             |          | material icons | override default icon                                                                                                                                                                                                                       |
|  renderToolbar   | (fullscreen) => ReactNode       |          |                | render something at the top (inside video), such as Title, Share button, More action                                                                                                                                                                                                                       |

**mode detail**

| Auto Fit                                                                                                        | Contain                                                                                                           |
| --------------------------------------------------------------------------------------------------------------- | --------------------------------------------------------------------------------------------------------------- |
| ![Auto Fit](https://user-images.githubusercontent.com/18292247/103472502-fb4b2e80-4dc0-11eb-96c3-88b5b059e1a3.jpg) | ![Contain](https://user-images.githubusercontent.com/18292247/103472503-fc7c5b80-4dc0-11eb-9f21-49242410a9ff.jpg) |

**customIcon example**

```tsx
import { Image } from 'react-native';

<YoutubePlayer
  mode="auto-fit"
  source={{
    uri: 'https://stream.mux.com/Tyu80069gbkJR2uIYlz2xARq8VOl4dLg3.m3u8',
  }}
  customIcon={{
    fullscreenIcon: (
      <Image
        src={require('../your-assets/fullscreen-icon.png')}
        style={{ width: 24, height: 24 }}
      />
    ),
    // exitFullscreenIcon: <SomeIcon />,
    // playIcon,
    // pauseIcon,
    // replayIcon,
    // forwardIcon,
    // refreshIcon,
    // volumeOffIcon,
    // volumeUpIcon,
  }}
/>;
```

### ScreenContainer

React Context that provide state of the video, useful when you want to do some fancy thing.
In this example, we need to disable scroll gesture in ScrollView while we are in fullscreen mode or dragging the thumb in the seeker.

```tsx
import { ScreenContainer, FacebookPlayer } from 'react-native-video-extension';

function App() {
  return (
    <ScreenContainer>
      {({ fullscreen, seeking }) => {
        return (
          <SafeAreaView
            style={{ flex: 1, backgroundColor: fullscreen ? '#000' : '#fff' }}
          >
            <ScrollView
              scrollEnabled={!fullscreen && !seeking}
              style={{ flex: 1 }}
              contentContainerStyle={{ flex: fullscreen ? 1 : 0 }}
            >
              {/* some header goes here */}
              <FacebookPlayer
                source={{
                  uri:
                    'https://stream.mux.com/Tyu80069gbkJR2uIYlz2xARq8VOl4dLg3.m3u8',
                }}
                mode="contain"
              />
              {/* some content goes here */}
            </ScrollView>
          </SafeAreaView>
        );
      }}
    </ScreenContainer>
  );
}
```

| name          | value                                                        | description                                                                    |
| ------------- | ------------------------------------------------------------ | ------------------------------------------------------------------------------ |
| fullscreen    | false \| "PORTRAIT" \| "LANDSCAPE-LEFT" \| "LANDSCAPE-RIGHT" | if not `false`, it provide the device orientation while in fullscreen mode     |
| seeking       | boolean                                                      | whether the thumb in seeker is dragging or not                                 |
| loading       | boolean                                                      | if `true`, video is not ready to play                                          |
| consoleHidden | boolean                                                      | if `true`, the controls is displaying (ex, play, forward, replay, seeker, ...) |
| isLandscape   | boolean                                                      | if `true`, the loaded video has naturalSize = "landscape"                      |

If you want to hook into the context in other component, use `useVideoCtx` like this.

```tsx
import { useEffect } from 'react';
import { View, Text } from 'react-native';
import { useVideoCtx } from 'react-native-video-extension';

// This component must be rendered inside ScreenContainer only!
function Awesome() {
  const { fullscreen } = useVideoCtx();
  useEffect(() => {
    if (fullscreen) {
      // Logger.send('fullscreen', 'uid)
      // just an example, I have no idea what I just wrote
    }
  }, [fullscreen]);
  return (
    <View>
      <Text>{fullscreen ? 'Wow' : ''}</Text>
    </View>
  );
}
```

## Integration

By default, this extension works without additional libraries, but the experience is not perfect.

#### Lock to Portrait

One thing to improve is to lock the screen to "Portrait" mode before entering fullscreen.

Install `react-native-orientation-locker` ([follow instruction here](https://github.com/wonday/react-native-orientation-locker#installation))

```tsx
// in app.tsx or the file that renders video
// ...other import
import {
  connectOrientationLib,
  YoutubePlayer,
} from 'react-native-video-extension';
import Orientation from 'react-native-orientation-locker';

connectOrientationLib(Orientation);

function Screen() {
  // ... code that render <YoutubePlayer />
}
```

That's it, try to restart the metro & simulator. navigate to the screen and press `⌘ + ←`. The video should enter fullscreen automatically.

#### Provide insets for some devices

Ex, iPhone 11 up has notch & spacing at the top, bottom or left,right. We can improve the experience by rendering the video in safe area.

Install `react-native-safe-area-context` ([follow instruction here](https://github.com/th3rdwave/react-native-safe-area-context#getting-started))

```tsx
// in app.tsx or the file that renders video
// ...other import
import { connectUseInsets, YoutubePlayer } from 'react-native-video-extension';
import { useSafeAreaInsets } from 'react-native-safe-area-context';

connectUseInsets(useSafeAreaInsets);

function Screen() {
  // ... code that render <YoutubePlayer />
}
```

Restart the metro & simulator, you should see the difference like the image below.

| Before                                                                                                          | After                                                                                                           |
| --------------------------------------------------------------------------------------------------------------- | --------------------------------------------------------------------------------------------------------------- |
| ![image](https://user-images.githubusercontent.com/18292247/103458307-4c174480-4d39-11eb-9e75-77f43024032e.png) | ![image](https://user-images.githubusercontent.com/18292247/103458311-546f7f80-4d39-11eb-8b82-ee9e9e8e82e1.png) |

## Incoming Feature

Take a look at [MVP project board](https://github.com/siriwatknp/react-native-video-extension/projects/1)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [0.1.3](https://github.com/siriwatknp/react-native-video-extension/compare/v0.1.2...v0.1.3) (2021-01-17)

### 0.1.2 (2021-01-03)


### 🚀 Features

* 🎸 support mux-data integration ([4a9afd3](https://github.com/siriwatknp/react-native-video-extension/commit/4a9afd35cae22e7d64d4a6dba38fa4584fb8b6d1))

### 0.1.1 (2021-01-03)

## 0.1.0 (2021-01-02)


### 🚀 Features

* 🎸 Public Release ([92edfcc](https://github.com/siriwatknp/react-native-video-extension/commit/92edfcc10e57760349289c4e5de8d7868d2e85eb))

### 0.0.2 (2021-01-02)

# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.


lib/utils.ts
import { VideoContext } from './ScreenContainer';
import { Inset } from './LayoutCalc';

const HOUR = 60 * 60;
const MINUTE = 60;
export const GUTTER_PERCENT = 5;
export const GUTTER_PX = 24;

export const isZeroInsets = (insets: Partial<Inset>) =>
  Object.entries(insets).every(([, value]) => value === 0);

const twoChar = (number: number): string =>
  Math.floor(number).toString().length === 1
    ? `0${Math.floor(number)}`
    : `${Math.floor(number)}`;

export const toTimeView = (sec: number) => {
  let secLeft = sec;
  const hr = Math.floor(sec / HOUR);
  secLeft = secLeft - hr * HOUR;
  const minute = Math.floor(secLeft / MINUTE);
  secLeft = secLeft - minute * MINUTE;
  if (hr && minute) {
    return `${hr}:${twoChar(minute)}:${twoChar(secLeft)}`;
  }
  if (hr && !minute) {
    return `${hr}:00:${twoChar(secLeft)}`;
  }
  if (minute) {
    return `${minute}:${twoChar(secLeft)}`;
  }
  return `0:${twoChar(secLeft)}`;
};

export const getSecondsToSeek = (
  duration: number,
  currentTime: number,
  seekerWidth: number,
  diffWidth: number,
  thumbRadius: number,
) => {
  const actualWidth = seekerWidth - 2 * thumbRadius;
  const result = currentTime + (diffWidth / actualWidth) * duration;
  if (result >= duration) return duration - 0.1;
  return result <= 0 ? 0.1 : result;
};

export const getThumbPosition = (
  duration: number,
  currentTime: number,
  seekerWidth: number,
  thumbRadius: number,
) => {
  if (!duration || !seekerWidth) return thumbRadius;
  const actualWidth = seekerWidth - 2 * thumbRadius;
  return (
    thumbRadius +
    (currentTime > duration ? 1 : currentTime / duration) * actualWidth
  );
};

export const calculateRotationDegree = (
  isLandscape: boolean,
  fullscreen?: VideoContext['fullscreen'],
) => {
  if (!isLandscape) return 0;
  if (fullscreen === 'LANDSCAPE-LEFT') return 90;
  if (fullscreen === 'LANDSCAPE-RIGHT') return -90;
  return 0;
};

export type AspectRatio = 'landscape' | 'portrait' | number;
export const getAspectRatio = (ratio: AspectRatio) => {
  if (typeof ratio === 'number') return ratio;
  if (ratio === 'portrait') return 3 / 4;
  return 16 / 9;
};
